### PR TITLE
8259319: Illegal package access when SunPKCS11 requires SunJCE's classes

### PR DESCRIPTION
**Pull request by martinuy**
As described in JDK-8259319 [1], this fix proposal is to set proper access permissions so the SunPKCS11 provider can create instances of SunJCE classes when a Security Manager is installed and the fallback scheme is used.

No regressions found in jdk/sun/security/pkcs11 tests category.

--
[1] - https://bugs.openjdk.java.net/browse/JDK-8259319

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259319](https://bugs.openjdk.java.net/browse/JDK-8259319): Illegal package access when SunPKCS11 requires SunJCE's classes

### Reviewers
 * [Valerie Peng](https://openjdk.java.net/census#valeriep) (@valeriepeng - **Reviewer**)
 * [Sean Mullan](https://openjdk.java.net/census#mullan) (@seanjmullan - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1961/head:pull/1961`
`$ git checkout pull/1961`

### CODE REVIEW

**Comment by valeriepeng** [comment]
Obscure bug, thanks for report and the fix. I will take a look.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java [hunk 1]
```diff
@@ -87,14 +87,19 @@ private static Provider getProvider(Provider p, String providerName,
...
         if (p == null) {
-            try {
-                @SuppressWarnings("deprecation")
-                Object o = Class.forName(className).newInstance();
-                p = (Provider)o;
-            } catch (Exception e) {
-                throw new ProviderException
-                        ("Could not find provider " + providerName, e);
-            }
+            p = AccessController.doPrivileged(
```
Comment (seanjmullan) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java:90]: Could you use the limited version of doPrivileged and only assert the permissions that are strictly necessary to instantiate a provider?

File: test/jdk/sun/security/pkcs11/KeyAgreement/IllegalPackageAccess.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        // Requires access to a SunJCE class that parses
+        // the encoded key.
+        kf.generatePublic(new X509EncodedKeySpec(encPubKey));
+
+        System.setSecurityManager(null);
+        Policy.setPolicy(policy);
+    }
+
+}
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/KeyAgreement/IllegalPackageAccess.java:96]: nit: add a newline here, to get rid of the red icon...

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java [hunk 3]
```diff
@@ -87,14 +87,19 @@ private static Provider getProvider(Provider p, String providerName,
...
+                        try {
+                            @SuppressWarnings("deprecation")
+                            Object o = Class.forName(className).newInstance();
+                            return (Provider) o;
+                        } catch (Exception e) {
+                            throw new ProviderException
+                                    ("Could not find provider " + providerName, e);
+                        }
+                    }
+                });
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java:102]: Sean's suggestion is to add additional arguments here, e.g. null, new RuntimePermission("accessClassInPackage." + <pkgName>).

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java (hunk 1 above)
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java]: Yes, makes sense. Thanks for your feedback.

File: test/jdk/sun/security/pkcs11/KeyAgreement/IllegalPackageAccess.java (hunk 2 above)
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/KeyAgreement/IllegalPackageAccess.java]: Yes, sure. Thanks for your feedback.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java (hunk 3 above)
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java]: Yes, replied to Sean above. Thanks.

**Comment by martinuy** [comment]
New proposal limiting the privileges in P11Util::getProvider method and adding a new line character at the end of the IllegalPackageAccess test file.

**Review by valeriepeng — APPROVED** [review]
Changes look good.

**Review by seanjmullan — APPROVED** [review]

**Pull request closed without merging**